File: 917-reverse-only-letters/917-reverse-only-letters.cpp
class Solution {
public:
    string reverseOnlyLetters(string s) {
        int i = 0, j = s.size()-1;
        bool fi = false, fj = false;
        while (i < j)
        {
            if (!fi && ((s[i] >= 'a' && s[i] <= 'z') || (s[i] >= 'A' && s[i] <='Z')))
                fi = true;
            if (!fj && ((s[j] >= 'a' && s[j] <= 'z') || (s[j] >= 'A' && s[j] <='Z')))
                fj = true;
            if (fi && fj)
            {
                swap(s[i], s[j]);
                fi = false, fj = false;
            }
            if (!fi)
                i++;
            if (!fj)
                j--;
        }
        return s;    
    }
};

File: 917-reverse-only-letters/917-reverse-only-letters.rb
# @param {String} s
# @return {String}
def reverse_only_letters(s)
    i, j = 0, s.length - 1
    fi, fj = false, false
    while i < j
       if !fi && ((s[i] >= 'a' && s[i] <= 'z') || (s[i] >= 'A' && s[i] <='Z'))
           fi = true
       end
        
        if !fj && ((s[j] >= 'a' && s[j] <= 'z') || (s[j] >= 'A' && s[j] <='Z'))
           fj = true 
        end
        
        if fi && fj
            s[i], s[j] = s[j], s[i]
            fi = false
            fj = false
        end
        if fi == false
            i += 1
        end
        if fj == false
            j -= 1
        end
    end
    return s
end